core/include/traccc/edm/details/device_container.hpp

```diff
@@ -0,0 +1,117 @@
+#include "traccc/definitions/qualifiers.hpp"
+#include "traccc/edm/details/container_base.hpp"
+
+// VecMem include(s).
+#include <vecmem/containers/device_vector.hpp>
+#include <vecmem/containers/jagged_device_vector.hpp>
+
+// System include(s).
+#include <cassert>
+#include <utility>
+
+namespace traccc {
+
+/// Host container describing objects in a given event
+template <typename header_t, typename item_t>
+class device_container
+    : public container_base<header_t, item_t, vecmem::device_vector,
+                            vecmem::jagged_device_vector, std::pair> {
+    public:
+    /// Base class type
+    using base_type = container_base<header_t, item_t, vecmem::device_vector,
+                                     vecmem::jagged_device_vector, std::pair>;
+
+    /// Inherit all of the base class's constructors
+    using base_type::base_type;
+
+    /**
+     * @brief Bounds-checking mutable element accessor.
+     */
+    TRACCC_HOST_DEVICE
+    typename base_type::element_view at(typename base_type::size_type i) {
+        assert(i < size());
+        return operator[](i);
+    }
+
+    /**
+     * @brief Bounds-checking immutable element accessor.
+     */
+    TRACCC_HOST_DEVICE
+    typename base_type::const_element_view at(
+        typename base_type::size_type i) const {
+        assert(i < size());
+        return operator[](i);
+    }
+
+    /**
+     * @brief Bounds-checking mutable item vector element accessor.
+     */
+    TRACCC_HOST_DEVICE
+    typename base_type::item_type& at(
+        const typename base_type::link_type& link) {
+        return this->m_items[link.first][link.second];
+    }
+
+    /**
+     * @brief Bounds-checking immutable item vector element accessor.
+     */
+    TRACCC_HOST_DEVICE
+    const typename base_type::item_type& at(
+        const typename base_type::link_type& link) const {
+        return this->m_items[link.first][link.second];
+    }
+
+    /**
+     * @brief Mutable element accessor.
+     */
+    TRACCC_HOST_DEVICE
+    typename base_type::element_view operator[](
+        typename base_type::size_type i) {
+        return {this->m_headers[i], this->m_items[i]};
+    }
+
+    /**
+     * @brief Immutable element accessor.
+     */
+    TRACCC_HOST_DEVICE
+    typename base_type::const_element_view operator[](
+        typename base_type::size_type i) const {
+        return {this->m_headers[i], this->m_items[i]};
+    }
+
+    /**
+     * @brief Return the size of the container.
+     *
+     * In principle, the size of the two internal vectors should always be
+     * equal, but we can assert this at runtime for debug builds.
+     */
+    TRACCC_HOST_DEVICE
+    typename base_type::size_type size(void) const {
+        return this->m_headers.size();
+    }
+
+    /**
+     * @breif Get number of items of jagged vector
+     */
+    TRACCC_HOST_DEVICE
+    uint64_t total_size() const {
+        uint64_t ret = 0;
+        for (auto& item : this->m_items) {
```

Comment:
I instead went all out, and used the concrete type name instead. :wink:

You see, `auto` in this case meant a constant type. This is one of the tricks that `auto` can do, that it can incorporate (amongst other things) `const` in some situations.

I also decided to move away from `uint64_t`, as I really just didn't like that. And I can honestly not see how 32-bits would not be enough for our use cases here...